### ISSUE TITLE
hdevtools does not find globally installed packages via stack such as QuickCheck 

### ISSUE THREAD

**Issue by dnivra** [hdevtools/hdevtools#30]
I am using stack 1.1.2 on Ubuntu 14.04. I have installed hdevtools, QuickCheck and checkers globally using stack. While I can import QuickCheck within stack ghci, hdevtools is not able to find it:

    Could not find module ‘Test.QuickCheck’
    Use -v to see a list of the files searched for.

**Comment by ch1bo** [comment]
Can you try a more recent version of `hdevtools`? Also, which stackage resolvers are you using? Either globally in `$HOME/.stack/global-project/stack.yaml` or within your project's `stack.yaml`

**Comment by dnivra** [comment]
Sorry about not including the version of hdevtools in the report! I'm using hdevtools 0.1.3.2 and lts-6.9. The latest LTS(6.10) has the same version as well.

**Comment by ch1bo** [comment]
Alright, can you try to checkout the `stack-ghc-libdir` branch, install that version using `stack install` and try this again? This version does not require execution `stack exec` anymore, so a direct invocation should find your compiler and packages correctly.

**Comment by dnivra** [comment]
Tried with `stack-ghc-libdir` branch. Same issue: hdevtools still doesn't find QuickCheck.

**Comment by ch1bo** [comment]
Okay can you call `hdevtools check <your file> --debug` ? Also make sure no (old) hdevtools instance is running. And you installed Quickqueck using stack?

**Comment by dnivra** [comment]
Yes I installed QuickCheck using stack. Here's the full output of the session:

```bash
❯ ls -a
.  ..  test.hs
❯ ps aux | grep -i hdevtools | grep -v grep
❯ cat test.hs
import Test.QuickCheck
❯ hdevtools check test.hs --debug
Cabal file: Nothing
Run from outside a project, using implicit global project config
Stack file: Just "/home/dnivra/.stack/global-project/stack.yaml"
Socket file: ".hdevtools.sock"
/home/dnivra/tmp/test/test.hs:1:8:
    Could not find module ‘Test.QuickCheck’
    Use -v to see a list of the files searched for.
```

**Comment by ch1bo** [comment]
So you have a dir/project without a cabal file? Can you help me out in determining the location of quickcheck package db in your case with `stack exec ghc-pkg -- list`?

**Comment by dnivra** [comment]
Yeah I suppose the global installation does not have a cabal file. I'm not sure how stack works. Here's the output of ghc-pkg -- list:

    ❯ stack exec ghc-pkg -- list
    Run from outside a project, using implicit global project config
    Using resolver: lts-6.9 from implicit global project's config file: /home/dnivra/.stack/global-project/stack.yaml
    /home/dnivra/.stack/programs/x86_64-linux/ghc-7.10.3/lib/ghc-7.10.3/package.conf.d
       Cabal-1.22.5.0
       array-0.5.1.0
       base-4.8.2.0
       bin-package-db-0.0.0.0
       binary-0.7.5.0
       bytestring-0.10.6.0
       containers-0.5.6.2
       deepseq-1.4.1.1
       directory-1.2.2.0
       filepath-1.4.0.0
       ghc-7.10.3
       ghc-prim-0.4.0.0
       haskeline-0.7.2.1
       hoopl-3.10.0.2
       hpc-0.6.0.2
       integer-gmp-1.0.0.0
       pretty-1.1.2.0
       process-1.2.3.0
       rts-1.0
       template-haskell-2.10.0.0
       terminfo-0.4.0.1
       time-1.5.0.1
       transformers-0.4.2.0
       unix-2.7.1.0
       xhtml-3000.2.1
    /home/dnivra/.stack/snapshots/x86_64-linux/lts-6.9/7.10.3/pkgdb
       Cabal-1.22.8.0
       HUnit-1.3.1.1
       MonadRandom-0.4.2.3
       QuickCheck-2.8.2
       StateVar-1.1.0.4
       ansi-terminal-0.6.2.3
       ansi-wl-pprint-0.6.7.3
       async-2.1.0
       base-orphans-0.5.4
       bifunctors-5.2
       cabal-helper-0.6.3.1
       checkers-0.4.5
       cmdargs-0.10.14
       comonad-4.2.7.2
       contravariant-1.4
       cpphs-1.20.1
       distributive-0.5.0.2
       djinn-ghc-0.0.2.3
       djinn-lib-0.0.1.2
       either-4.4.1.1
       exceptions-0.8.3
       extra-1.4.10
       fclabels-2.0.3
       free-4.12.4
       ghc-mod-5.5.0.0
       ghc-paths-0.1.0.9
       ghc-syb-utils-0.2.3
       hashable-1.2.4.0
       haskell-src-exts-1.17.1
       hlint-1.9.35
       hscolour-1.24.1
       hspec-2.2.3
       hspec-core-2.2.3
       hspec-discover-2.2.3
       hspec-expectations-0.7.2
       mmorph-1.0.6
       monad-control-1.0.1.0
       monad-journal-0.7.2
       mtl-2.2.1
       network-2.6.2.1
       old-locale-1.0.0.7
       old-time-1.1.0.3
       optparse-applicative-0.12.1.0
       pipes-4.1.9
       polyparse-1.12
       prelude-extras-0.4.0.3
       primitive-0.6.1.0
       profunctors-5.2
       quickcheck-io-0.1.3
       random-1.1
       refact-0.3.0.2
       safe-0.3.9
       semigroupoids-5.0.1
       semigroups-0.18.1
       setenv-0.1.1.3
       split-0.2.3.1
       stm-2.4.4.1
       syb-0.6
       tagged-0.8.4
       temporary-1.2.0.4
       text-1.2.2.1
       tf-random-0.5
       transformers-base-0.4.4
       transformers-compat-0.4.0.4
       uniplate-1.6.12
       unordered-containers-0.2.7.1
       utf8-string-1.0.1.1
       void-0.7.1
    /home/dnivra/.stack/global-project/.stack-work/install/x86_64-linux/lts-6.9/7.10.3/pkgdb

**Comment by ch1bo** [comment]
I think I have a good idea how to reproduce your situation.

Honestly I never tried to work without a cabal file as project description. Do you use hpack or something else?

I think your isse goes away as soon hdevtools finds a cabal file containing the quickqueck dependancy. Is this an option for you?

**Comment by dnivra** [comment]
I don't know if hpack is used. I simply run `stack setup` and then `stack install` to install packages globally.

I don't see why a cabal file is required. stack seems to work fine with globally installed packages. As per my limited understanding, it doesn't make sense to have a cabal configuration file for the global installation since there's no project, dependencies etc.

**Comment by uuhan** [comment]
Please DO NOT rely on a .cabal file just for  writing some haskell code without cabal. (hdevtools 0.1.4.0).
**stack** supplies enough information for what packages (in snapshots) I am going to search for.

**Comment by ch1bo** [comment]
Do I understand you correctly in thath you suggest all available packages of a certain snapshot should be loaded when typechecking in a project without cabal file?

**Comment by uuhan** [comment]
@ch1bo It seems that I make some missunderstanding when I have hdevtools binary both in ~/.cabal/bin (installed by cabal) and ~/.local/bin (installed by stack). 
It works just fine when I use the one in ~/.cabal/bin to  write haskell programs without <project>.cabal file. : )

**Comment by ch1bo** [comment]
Okay perfect.

Btw, the one installed by cabal comes directly vom hackage and is most likely a newer version. Because wenn you install via stack without a project stack.yaml, the resolver/snapshot from the global stack.yaml in ~/.stack/global/stack.yaml (or similar) is used - and this has to be kept up to date to get the newest hdevtools releases.

**Comment by stschiff** [comment]
I think there is still a problem here. Stack introduced the concept of [stack scripts](https://github.com/commercialhaskell/stack/wiki/Script-interpreter), by putting two lines on top of your haskell source file, e.g.:

    #!/usr/bin/env stack
    -- stack runghc -- package turtle

Now, when executing this file, stack invokes the global stack installation and loads any packages it needs. This is tremdeously useful e.g. to write [turtle scripts](https://hackage.haskell.org/package/turtle). It takes out a lot of boilerplate if I don't have to define a .cabal file any time I'd like to write a script like this.

Note that you don't even have to give `package turtle` in the second header line. You could just run `stack install turtle` outside of a project once, and then it will be found automatically by `stack runghc`.

The problem is that any of those scripts need to at least `import turtle`. Running the script is not a problem because stack will install turtle in the global package-repo on the fly for us. But when running hdevtools, we get an error `Failed to load interface for ‘Turtle’`.

I do not expect hdevtools to actually parse the shebang and header lines of the script, but it would be nice if it was smart enough to check the global .stack file and load all necessary globally installed packages somehow.

**Comment by ch1bo** [comment]
>  it would be nice if it was smart enough to check the global .stack file and load all necessary globally installed packages somehow.

Good idea, this issue should be solvable

**Comment by EdmundsEcho** [comment]
Great workflow tool.

When running both `stack-ghc-mod` and `hdevtools`, it's clear how much `hdevtools` is unique in the inability to use the globally installed packages.  I remove `hdevtools` from the list of linters and `stack-ghc-mod` works as expected - no unnecessary errors.  Is this a bug fix something that remains a priority?
